Add support to rename clusters and cluster replicas.

antiguru opened the pull request:
Add support to rename cluster and cluster replicas.

### Motivation

As part of the managed clusters effort (#19547), we are improving the interface to work with clusters, and one part of this is to add the ability to rename clusters and their replicas. This PR adds the functionality to do that, specifically it enables the following SQL syntax:

```sql
-- Assume cluster `foo` exists
ALTER CLUSTER foo RENAME TO bar;
-- Cluster `foo` is now known as `bar`.
ALTER CLUSTER REPLICA bar.alice RENAME TO bob
-- Cluster replica `bar.alice` is now known as `bar.bob`.
```

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Add support for `ALTER CLUSTER [REPLICA] ... RENAME TO ...`.

antiguru commented:
I marked this PR as ready for review, but I'm certain it still has wide holes that I still have to learn about! I left some `TODO` questions in the source code. I'm particularly interested in whether the interactions with the audit log are correct (in other places we have a `Rename` type, but that doesn't apply to cluster/replicas), and whether the whole structure makes sense.

benesch commented on src/adapter/src/catalog.rs:
```diff
@@ -5995,6 +6028,79 @@ impl Catalog {
+                        audit_events,
+                        EventType::Alter,
+                        ObjectType::Cluster,
+                        // TODO: Is this the correct EventDetails variant?
```
I think you want `RenameItemV1`.

Can you also add a test to the audit_log.slt file that exercises the audit log output for cluster and cluster replica renaming?

benesch replied:
Ah good point. `RenameCluster` sounds right. I don't think there's enough generality to pursue `RenameObject` and this approach is only a little bit more typing.

benesch commented on src/adapter/src/catalog.rs:
```diff
@@ -5995,6 +6028,79 @@ impl Catalog {
+                        audit_events,
+                        EventType::Alter,
+                        ObjectType::ClusterReplica,
+                        // TODO: Is this the correct EventDetails variant?
+                        EventDetails::IdNameV1(mz_audit_log::IdNameV1 {
+                            // TODO: What should this `id` be?
```
Think you need to introduce a new `RenameClusterReplicaV1` type here that has separate `cluster_id` and `replica_id` fields, to parallel `CreateClusterReplicaV1`.

benesch commented on src/adapter/src/catalog/storage.rs:
```diff
@@ -992,6 +993,24 @@ impl<'a> Transaction<'a> {
         )
     }
 
+    pub fn rename_user_cluster(
+        &mut self,
+        cluster_id: ClusterId,
+        cluster_name: &str,
+        cluster_to_name: &str,
+    ) -> Result<(), Error> {
+        self.rename_cluster(cluster_id, cluster_name, cluster_to_name)
+    }
+
+    pub fn rename_user_cluster_replica(
+        &mut self,
+        replica_id: ReplicaId,
+        replica_name: &QualifiedReplica,
+        replica_to_name: &str,
+    ) -> Result<(), Error> {
+        self.rename_cluster_replica(replica_id, replica_name, replica_to_name)
+    }
```
These seem a bit weird, in that they don't differ from the non `user` variants, and don't limit cluster/replica IDs to just the user variants. Can we just not have them, and have callers call `rename_cluster[_replica]` directly?

antiguru replied:
Removed.

benesch commented on src/adapter/src/catalog.rs:
```diff
@@ -973,6 +973,15 @@ impl CatalogState {
     }
 
+    fn rename_cluster(&mut self, id: ClusterId, to_name: String) {
+        let cluster = self.clusters_by_id.get_mut(&id).expect("Missing cluster");
```
We've got a purpose-built helper for this that includes the missing cluster ID in the panic message.

```suggestion
        let cluster = self.get_cluster_mut(&id);
```

benesch commented on src/adapter/src/catalog.rs:
```diff
@@ -1008,6 +1017,30 @@ impl CatalogState {
+        replica_id: ReplicaId,
+        to_name: String,
+    ) {
+        let cluster = self
+            .clusters_by_id
+            .get_mut(&cluster_id)
+            .expect("Missing cluster");
+        let replica = cluster
+            .replicas_by_id
+            .get_mut(&replica_id)
+            .expect("Missing cluster");
```
```suggestion
        let replica = self.get_cluster_replica_mut(cluster_id, replica_id);
```

You'll need to add this helper, but I think it's worth it if you do so with the same panic message that `get_cluster_replica` uses. We have every other variant: `try_get_cluster_replica`, `get_cluster_replica`, and `try_get_cluster_replica_mut`.

antiguru replied:
Thanks, fixed.

benesch reviewed (commented):
Wow, that was fast! Left some line notes within but generally looks great. We should have someone from surfaces take a look too though.

Testing wise, we should have team QA take a look. I'll request a review from @philip-stoev so he can delegate appropriately. The one missing test that jumped out to me, which I left a line note about, was adding rename operations to audit_log.slt, so that we can see what this looks like in the audit log.

antiguru commented:
I addressed the comments. I'm not yet handling no-op renames in a special way, which might lead to confusing entries in the audit log. Should this be handled, or is it fine? Renaming items seems to not handle this case.

jkosh44 commented on test/sqllogictest/cluster.slt:
Can we add the following tests:
  - one that uses `IF EXISTS`.
  - one without `IF EXISTS` that references a non-existent object.
  - one that uses the wrong object type, for example `ALTER CLUSTER REPLICA default RENAME TO bar`

antiguru replied:
Done.

jkosh44 commented on src/stash/protos/objects_v17.proto:
@parker-timmerman Can you comment on if this meets the criteria of needing a new snapshot? 

As a side note we should probably add a checklist item to PR template for taking a snapshot of the objects file.

ParkMyCar replied:
While this change doesn't itself technically require a snapshot, we should include one to stay in the habit of always snapshoting our changes.

@antiguru you should also bump the [STASH_VERSION](https://github.com/MaterializeInc/materialize/blob/main/src/stash/src/lib.rs#L105) and then [here](https://github.com/MaterializeInc/materialize/blob/8a7dbc655df15f6d38b2ed79a4280d729ccee8f2/src/stash/src/postgres.rs#L948) include a no-op migration, [example](https://github.com/MaterializeInc/materialize/blob/8a7dbc655df15f6d38b2ed79a4280d729ccee8f2/src/stash/src/upgrade/v13_to_v14.rs).

> As a side note we should probably add a checklist item to PR template for taking a snapshot of the objects file.

This is a good idea, I'll also add a test to make sure that `STASH_VERSION` stays in sync with how many snaphots we have.

antiguru replied:
Thanks, added!

jkosh44 commented on src/adapter/src/catalog.rs:
```diff
@@ -5995,6 +6036,81 @@ impl Catalog {
+                    name,
+                    to_name,
+                } => {
+                    if cluster_id.is_system() {
+                        return Err(AdapterError::Catalog(Error::new(
+                            ErrorKind::ReadOnlyCluster(name.cluster.into_string()),
+                        )));
+                    }
```
Or did that change never land?

benesch replied:
It's still in progress, looks like: https://github.com/MaterializeInc/materialize/pull/18636

jkosh44 commented:
> I'm not yet handling no-op renames in a special way, which might lead to confusing entries in the audit log. Should this be handled, or is it fine? Renaming items seems to not handle this case.

If it's consistent with renaming items then I think it's fine for this PR, but we should open up an issue to convert those cases into a no-op.

benesch reviewed (commented):
I'm at the limit of what I'm capable of reviewing, but everything I looked at LGTM. Deferring full approval to @jkosh44.

Thanks again for turning this around so quickly! 

Oh, one last thing: docs! We should make sure to update the reference docs for the existence of `ALTER CLUSTER` and `ALTER CLUSTER REPLICA`.

antiguru commented on src/adapter/src/rbac.rs:
```diff
@@ -351,6 +354,9 @@ fn generate_required_plan_attribute(plan: &Plan) -> Vec<Attribute> {
         }
+        Plan::AlterClusterRename(_) | Plan::AlterClusterReplicaRename(_) => {
+            vec![Attribute::CreateCluster]
+        }
```
Removed, thanks.

antiguru commented on src/adapter/src/rbac.rs:
```diff
@@ -1011,6 +1021,12 @@ fn generate_required_privileges(
         }
+        Plan::AlterClusterRename(AlterClusterRenamePlan { id, .. }) => {
+            vec![(id.into(), AclMode::CREATE, role_id)]
+        }
```
Removed, thanks.

def- reviewed (commented):
I'll take a look from QA side.

def- commented on src/adapter/src/catalog/storage.rs:
```diff
@@ -1053,6 +1054,50 @@ impl<'a> Transaction<'a> {
+            }
+        })? {
+            1 => Ok(()),
+            _ => Err(SqlCatalogError::UnknownCluster(cluster_name.to_string()).into()),
```
Untested error case

antiguru replied:
This case can only happen when the catalog state is out-of-sync, which is not a case I can test. I'm not sure that there is any benefit in returning this error over panicking? Maybe @jkosh44 has an opinion here.

def- commented on src/adapter/src/catalog/storage.rs:
```diff
@@ -1053,6 +1054,50 @@ impl<'a> Transaction<'a> {
+            }
+        })? {
+            1 => Ok(()),
+            _ => Err(SqlCatalogError::UnknownClusterReplica(replica_name.to_string()).into()),
```
Untested error case

antiguru replied:
Dito.

def- commented on src/adapter/src/rbac.rs:
```diff
@@ -560,6 +565,10 @@ fn generate_required_ownership(plan: &Plan) -> Vec<ObjectId> {
             .unwrap_or_default(),
         // Do not need ownership of descendant objects.
         Plan::DropObjects(plan) => plan.referenced_ids.clone(),
+        Plan::AlterClusterRename(plan) => vec![ObjectId::Cluster(plan.id)],
+        Plan::AlterClusterReplicaRename(plan) => {
+            vec![ObjectId::ClusterReplica((plan.cluster_id, plan.replica_id))]
```
Could add RBAC tests to check if renaming them works/fails as expected. (I'm kind of wondering that this shows up as uncovered, we should have RBAC enabled by default in CI)

antiguru replied:
That's a good point. I think we're missing a discussion about which privileges would be required to rename clusters/replicas. I tried to add tests, but the crux is that I don't know whether renaming should require a specific privilege or not. @jkosh44 let's talk about this once you're back!

jkosh44 replied:
Renaming an object (and all other alter's) is the equivalent of dropping that object and then re-creating it from a privilege point of view. That means for clusters you need:
- Ownership of the cluster.
- The `CREATECLUSTER` attribute.
For cluster replicas you need:
- Ownership of the replica.
- `CREATE` privileges on the containing cluster.

jkosh44 replied:
Wait sorry, I think I'm mistaken here. It only requires Ownership for both. Sorry about that.

def- commented on src/adapter/src/rbac.rs:
```diff
@@ -1011,6 +1020,9 @@ fn generate_required_privileges(
             }
             privileges
         }
+        Plan::AlterClusterReplicaRename(AlterClusterReplicaRenamePlan { cluster_id, .. }) => {
```
Another uncovered one?

antiguru replied:
Removed, was a leftover.

def- reviewed (commented):
As @jkosh44 already mentioned, some of the error cases are not tested. I checked coverage via https://buildkite.com/materialize/coverage/builds/126 and marked the interesting spots inline.

jseldess commented on doc/user/content/sql/alter-rename.md:
````diff
@@ -72,6 +72,8 @@ You cannot rename items if:
     ON v1.a = v2.a
     ```
 
+- Rename clusters and replicas if they are system clusters. For example, the `mz_introspection` cluster cannot be renamed.
````
Tweak to work with the list intro (`Your cannot rename items if:`):

```suggestion
- They are system clusters, such as `mz_system` and `mz_introspection`, or replicas in system clusters.
```

jseldess reviewed (approved):
Docs LGTM, with a suggestion.

def- reviewed (approved):
I added platform-checks, nightly is fine with it (the upgrade-across-4-versions failure is unrelated and already fixed): https://buildkite.com/materialize/nightlies/builds/2519

jkosh44 reviewed (approved):
LGTM

jkosh44 commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -4214,6 +4246,46 @@ pub fn plan_alter_object_rename(
+        Some(entry) => Ok(Plan::AlterClusterRename(AlterClusterRenamePlan {
+            id: entry.id(),
+            name: entry.name().to_string(),
+            to_name: to_name.into_string(),
```
I just realized, I think we need to normalize this name like we do for replica names. Unless there's a reason that I'm unaware of.

benesch replied:
`into_string` and `normalize::ident` do the same thing. It's subtle though and `normalize::ident` would be better. Also `normalize::ident` should use the new `into_string` function` instead of `as_str().to_string()`.

jkosh44 replied:
Ah got it. It might even be better to just move this call up a level into `plan_alter_object_rename` so they're all handled the same and it might help avoid potential confusion.

jkosh44 commented:
> > I'm not yet handling no-op renames in a special way, which might lead to confusing entries in the audit log. Should this be handled, or is it fine? Renaming items seems to not handle this case.
> 
> If it's consistent with renaming items then I think it's fine for this PR, but we should open up an issue to convert those cases into a no-op.

FYI: https://github.com/MaterializeInc/materialize/issues/19712

antiguru commented:
@mjibson LGTM'ed via Slack.

benesch commented:
Sweet! Thanks for getting this all in order, Moritz!

On Mon, Jun 5, 2023 at 10:51 PM Moritz Hoffmann ***@***.***>
wrote:

> Merged #19642 <https://github.com/MaterializeInc/materialize/pull/19642>
> into main.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MaterializeInc/materialize/pull/19642#event-9440653746>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGXSIEPDJ6C5SMDDPR4KT3XJ2LL7ANCNFSM6AAAAAAYXIAK2A>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>
